### PR TITLE
Client

### PR DESCRIPTION
**Pull request by airhorns**
This introduces the first bits of a CM API client to this gem. This is a PR on an empty repo.
- Uses the `her` gem https://github.com/remiprev/her to wrap the CM API with a few Faraday middlewares
- Implements Host, Cluster, and Command objects that represent those entities in the CM API
- Creates a `BaseResource` class each resource can inherit from to leverage common functionality like executing commands or fetching properties from other routes on the resource
- Adds a test suite with fixtures based on our CM deploy but neutered to not have any sensitive information
- Doesn't do anything else, that will come later!

### CODE REVIEW

**Comment by airhorns** [comment]
@udnay @dklassen

File: lib/cloudera-manager/host.rb [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+    class << self
+      def find_by_hostname(hostname)
+        if host = all.to_a.find {|host| host.hostname == hostname }
+          return host
+        else
+          raise HostNotFoundException.new("Hostname #{hostname} could not be found remotely")
+        end
+      end
+
+      def hadoopy?(hostname)
```
Comment (udnay) [inline lib/cloudera-manager/host.rb:16]: I woulda gone with `haderp` :trollface:

File: test/cloudera-manager/host_test.rb [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+  end
+
+  def test_getting_a_host_retrieves_a_host
+    @host = ClouderaManager::Host.find('4f8ebfcb-20c9-47f1-bd0e-cd3f1d8c891e')
+    assert_equal 'hadoop-util1.chi.shopify.com', @host.hostname
+  end
+
+  def test_retrieving_hosts_by_hostname
+    @host = ClouderaManager::Host.find_by_hostname('hadoop-util1.chi.shopify.com')
+    assert_equal 'hadoop-util1.chi.shopify.com', @host.hostname
```
Comment (udnay) [inline test/cloudera-manager/host_test.rb:16]: Maybe assert the id as well. Also `hadoop-util1` no longer exists

**Comment by udnay** [comment]
This looks like an awesome start. `her` seems really fucking cool as well.

File: lib/cloudera-manager-client.rb [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+    @api ||= Her::API.new
+  end
+
+  def self.setup(options, &blk)
+    username = options.delete(:username)
+    password = options.delete(:password)
+
+    api.setup(options) do |conn|
+      conn.use Her::Middleware::AcceptJSON
+      conn.use ClouderaManager::Middleware::URLPrefix, '/api/v10'
```
Comment (dklassen) [inline lib/cloudera-manager-client.rb:17]: *nitpick. Is there any use to having this customizable? If you are going to hard code it here why not hard code it in the middleware?

Comment (airhorns) [inline lib/cloudera-manager-client.rb:17]: I guess I liked the prefix being on the outside here such that you could see it, and I copy pasted the url prefix middleware from elsewhere :) I kind of prefer it being obvious like this.

**Comment by airhorns** [comment]
Are you guys cool if I ship this and https://github.com/Shopify/spy/pull/739 ?

**Comment by udnay** [comment]
Yep, sounds good.

**Pull request merged**